web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js

```diff
@@ -1,18 +1,295 @@
 import {
+	afterAll,
 	afterEach,
 	describe,
 	expect,
-	it
+	it,
+	vi
 } from "vitest";
-import { cleanup, render } from "@testing-library/svelte";
+import { cleanup, fireEvent, render } from "@testing-library/svelte";
+import { generateMnemonic } from "bip39";
+import * as appNavigation from "$app/navigation";
+import { Wallet } from "@dusk-network/dusk-wallet-js";
+import { setKey } from "lamb";
+import { addresses } from "$lib/mock-data";
+import { settingsStore, walletStore } from "$lib/stores";
+import { encryptMnemonic, getSeedFromMnemonic } from "$lib/wallet";
+import loginInfoStorage from "$lib/services/loginInfoStorage";
+import * as walletService from "$lib/services/wallet";
 import Restore from "../+page.svelte";
 
-describe("Restore", () => {
-	afterEach(cleanup);
+/**
+ * @param {HTMLElement} input
+ * @param {*} value
+ * @returns {Promise<boolean>}
+ */
+const fireInput = (input, value) => fireEvent.input(input, { target: { value } });
+
+/** @param {HTMLElement} element */
+function asInput (element) {
+	// eslint-disable-next-line no-extra-parens
+	return /** @type {HTMLInputElement} */ (element);
+}
+
+describe("Restore", async () => {
+	const walletGetPsksSpy = vi.spyOn(Wallet.prototype, "getPsks").mockResolvedValue(addresses);
+	const mnemonic = generateMnemonic();
+	const invalidMnemonic = "dad dad dad dad dad dad dad dad dad dad dad dad";
+	const pwd = "passwordpassword";
+	const loginInfo = await encryptMnemonic(mnemonic, pwd);
+	const seed = getSeedFromMnemonic(mnemonic);
+	const userId = (await new Wallet(seed).getPsks())[0];
+	const getWalletSpy = vi.spyOn(walletService, "getWallet");
+	const gotoSpy = vi.spyOn(appNavigation, "goto");
+	const settingsResetSpy = vi.spyOn(settingsStore, "reset");
+	const originalNavigator = { ...window.navigator };
+	const clearAndInitSpy = vi.spyOn(walletStore, "clearLocalDataAndInit");
+
+	afterEach(async () => {
+		cleanup();
+		clearAndInitSpy.mockClear();
+		getWalletSpy.mockClear();
+		gotoSpy.mockClear();
+		settingsStore.reset();
+		settingsResetSpy.mockClear();
+		walletGetPsksSpy.mockClear();
+		walletStore.reset();
+		Object.defineProperty(window, "navigator", {
```

Comment:
You don't need to do this after each test: you can set it up before with something alone the lines of:

```js
const readTextMock = vi.fn().mockResolvedValue(mnemonic);

Object.assign(window.navigator, { clipboard: { readText: readTextMock } } );
```

Then in the `afterEach` block you just clear the mock.

Reply:
Have you tested this approach and got the same results?

Reply:
even better results! (the paste button is not in the snapshot now, for example)

In this case you'll need to use `act` to have the mock promise settled though (or await a `tick`).

Reply:
> even better results! (the paste button is not in the snapshot now, for example)
> 
> In this case you'll need to use `act` to have the mock promise settled though (or await a `tick`).

Why should the paste button not be in the snapshot? It is part of the markup and should be shown.

Reply:
> Why should the paste button not be in the snapshot? It is part of the markup and should be shown.

Yes, but it isn't there in your current snaphshot. With the change above it'll be there from the start.